src/node_jsvmapi.cc

```diff
@@ -465,22 +465,109 @@ namespace v8impl {
 
 }  // end of namespace v8impl
 
-// We define a macro here because, with full error handling, we may want to
-// perform additional actions at the top of our public APIs, such as bail,
-// returning an error code at the very top of each public API in the case that
-// there's already an exception pending in v8impl::TryCatch::lastException
-#define NAPI_PREAMBLE(e) \
-  v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)))
+#define RETURN_STATUS_IF_FAIL(condition, status)                        \
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      return napi_set_last_error_code((status));                        \
+    }                                                                   \
+  } while(0)
+
+#define CHECK_ARG(arg)                                                  \
+  RETURN_STATUS_IF_FAIL((arg), napi_invalid_arg)
+
+#define CHECK_MAYBE_EMPTY(maybe, status)                                \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsEmpty()), (status))
+
+#define CHECK_MAYBE_NOTHING(maybe, status)                              \
+  RETURN_STATUS_IF_FAIL(!((maybe).IsNothing()), (status))
+
+#define NAPI_PREAMBLE(e)                                                \
+  do {                                                                  \
+    RETURN_STATUS_IF_FAIL(v8impl::TryCatch::lastException().IsEmpty(),  \
+      napi_pending_exception);                                          \
+    napi_clear_last_error_code();                                       \
+    v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)));         \
+  } while(0)
+
+#define CHECK_TO_OBJECT(result, o)                                      \
```

Comment:
We could probably use a `ToString()` flavor of this also. Or perhaps use preprocessor token pasting to make macro that works with any type:
```
#define CHECK_TO_X(type, result, o)  \
    ...
    auto maybe = v8impl::V8LocalValueFromJsValue((o))->To##type(context);
    ...

#define CHECK_TO_STRING(result, o)  CHECK_TO_X(String, result, o)
```